# PipelineWatch-NG — Module 3: ML Anomaly Detection
### Trans Niger Pipeline corridor · Niger Delta, Nigeria

**Run Modules 1 and 2 first**, then open this notebook.

| Step | Method | Input |
|------|--------|-------|
| Risk scorer | XGBoost | m2_feature_table.csv |
| Anomaly detector | Isolation Forest | m2_feature_table.csv |
| Feature importance | SHAP values | XGBoost model |
| Output | Risk-scored CSV + model files | Module 4 dashboard |

---

## Cell 1 — Imports

In [1]:
import os
import json
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import warnings
warnings.filterwarnings("ignore")
from datetime import datetime
from sklearn.ensemble import IsolationForest
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import classification_report, confusion_matrix
import xgboost as xgb
import joblib

CACHE_DIR  = "../data/cached"
OUTPUT_DIR = "../outputs"
MODEL_DIR  = "../data/models"
os.makedirs(MODEL_DIR, exist_ok=True)
os.makedirs(OUTPUT_DIR, exist_ok=True)

print("Imports OK.")
print("XGBoost version: " + xgb.__version__)


Imports OK.
XGBoost version: 3.2.0


## Cell 2 — Load feature table from Module 2

In [2]:
feat_path = os.path.join(CACHE_DIR, "m2_feature_table.csv")
if not os.path.exists(feat_path):
    print("ERROR: m2_feature_table.csv not found — run Module 2 first")
else:
    df = pd.read_csv(feat_path)
    print("Feature table loaded: " + str(df.shape[0]) + " rows x " + str(df.shape[1]) + " columns")
    print("Columns: " + str(df.columns.tolist()))
    print()
    print(df.describe().round(3))


Feature table loaded: 270 rows x 13 columns
Columns: ['NDMI', 'NDVI', 'NDVI_change', 'NDWI', 'SAR_change_dB', 'VH', 'VV', 'chronic_spill_mask', 'dark_spot_magnitude', 'dark_spot_mask', 'new_spill_mask', 'lon', 'lat']

          NDMI     NDVI  NDVI_change     NDWI  SAR_change_dB       VH  \
count  270.000  270.000      270.000  270.000        270.000  270.000   
mean     0.089    0.493        0.033   -0.469         -0.023  -14.543   
std      0.082    0.172        0.102    0.149          0.417    1.258   
min     -0.162   -0.242       -0.321   -0.681         -1.487  -23.754   
25%      0.041    0.396       -0.036   -0.570         -0.253  -15.007   
50%      0.088    0.521        0.041   -0.496         -0.020  -14.552   
75%      0.143    0.611        0.108   -0.392          0.166  -13.984   
max      0.309    0.768        0.272    0.253          2.831   -9.252   

            VV  chronic_spill_mask  dark_spot_magnitude  dark_spot_mask  \
count  270.000             270.000              2

## Cell 3 — Feature engineering and label creation

We have no ground-truth labels (no confirmed theft incident database for the TNP corridor).
This is the reality of most real-world environmental monitoring problems.

**Strategy: rule-based weak labelling**
We use domain knowledge to create proxy labels from signal combinations:
-  risk: SAR dark spot AND (fire hotspot nearby OR strong SAR change)
-  risk: any single strong signal above threshold
-  risk: all signals near background levels

This is a standard technique in remote sensing ML — the model then learns
to generalise beyond the hard threshold rules.

In [3]:
FEATURE_COLS = [
    'VV', 'VH', 'dark_spot_mask', 'dark_spot_magnitude',
    'SAR_change_dB', 'new_spill_mask', 'chronic_spill_mask',
]
optional_cols = ['NDVI', 'NDWI', 'NDMI', 'NDVI_change']
for col in optional_cols:
    if col in df.columns:
        FEATURE_COLS.append(col)

print('Feature columns: ' + str(FEATURE_COLS))

df_clean = df[FEATURE_COLS + ['lat', 'lon']].dropna().copy()
print('Clean rows: ' + str(len(df_clean)))

def assign_risk_label(row):
    """
    Relaxed thresholds — tuned to the actual data distribution.
    Uses percentile-based dark spot magnitude instead of hard cutoffs.
    """
    sar_dark   = row['dark_spot_mask'] == 1
    sar_change = row['SAR_change_dB'] < -0.5   # relaxed from -3.0
    chronic    = row['chronic_spill_mask'] == 1
    dark_mag   = row['dark_spot_magnitude'] > 1.0  # relaxed from 2.0

    ndvi_drop = False
    if 'NDVI_change' in row.index:
        ndvi_drop = row['NDVI_change'] < -0.10  # relaxed from -0.15

    vv_low = row['VV'] < -10.0  # very low backscatter

    if (sar_dark or chronic) and (dark_mag or ndvi_drop):
        return 2  # HIGH
    elif sar_change or vv_low or ndvi_drop or dark_mag:
        return 1  # MEDIUM
    else:
        return 0  # LOW

df_clean['risk_label'] = df_clean.apply(assign_risk_label, axis=1)

label_counts = df_clean['risk_label'].value_counts().sort_index()
print()
print('Risk label distribution:')
for label, count in label_counts.items():
    name = {0: 'LOW', 1: 'MEDIUM', 2: 'HIGH'}[label]
    pct  = round(100 * count / len(df_clean), 1)
    print('  ' + name + ': ' + str(count) + ' samples  (' + str(pct) + '%)')

Feature columns: ['VV', 'VH', 'dark_spot_mask', 'dark_spot_magnitude', 'SAR_change_dB', 'new_spill_mask', 'chronic_spill_mask', 'NDVI', 'NDWI', 'NDMI', 'NDVI_change']
Clean rows: 270

Risk label distribution:
  LOW: 214 samples  (79.3%)
  MEDIUM: 53 samples  (19.6%)
  HIGH: 3 samples  (1.1%)


## Cell 4 — XGBoost risk scorer

XGBoost is well-suited here because:
- Handles mixed feature types (binary masks + continuous values)
- Robust to class imbalance (most pixels are LOW risk)
- Fast training on small datasets (270 rows)
- Native feature importance — tells us which sensor matters most

In [4]:
X = df_clean[FEATURE_COLS].values
y = df_clean["risk_label"].values

# Scale features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Train/test split — stratified to preserve class distribution
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42
)

print("Training set : " + str(len(X_train)) + " samples")
print("Test set     : " + str(len(X_test))  + " samples")
print()

# XGBoost classifier
# scale_pos_weight handles class imbalance
n_low    = (y_train == 0).sum()
n_medium = (y_train == 1).sum()
n_high   = (y_train == 2).sum()

xgb_model = xgb.XGBClassifier(
    n_estimators=200,
    max_depth=4,
    learning_rate=0.1,
    subsample=0.8,
    colsample_bytree=0.8,
    use_label_encoder=False,
    eval_metric="mlogloss",
    random_state=42,
    verbosity=0,
)

xgb_model.fit(
    X_train, y_train,
    eval_set=[(X_test, y_test)],
    verbose=False,
)

# Evaluate
y_pred = xgb_model.predict(X_test)
print("Classification report:")
print(classification_report(y_test, y_pred,
      target_names=["LOW","MEDIUM","HIGH"],
      zero_division=0))

# Cross-validation
cv_scores = cross_val_score(xgb_model, X_scaled, y, cv=5, scoring="accuracy")
print("5-fold CV accuracy: " + str(round(cv_scores.mean(), 3))
      + " +/- " + str(round(cv_scores.std(), 3)))


Training set : 216 samples
Test set     : 54 samples

Classification report:
              precision    recall  f1-score   support

         LOW       0.95      1.00      0.98        42
      MEDIUM       1.00      0.91      0.95        11
        HIGH       0.00      0.00      0.00         1

    accuracy                           0.96        54
   macro avg       0.65      0.64      0.64        54
weighted avg       0.95      0.96      0.95        54

5-fold CV accuracy: 0.981 +/- 0.017


## Cell 5 — Feature importance
Shows which sensor drives the risk score most strongly.

In [5]:
importances = xgb_model.feature_importances_
feat_imp = pd.DataFrame({
    'feature':    FEATURE_COLS,
    'importance': importances
}).sort_values('importance', ascending=False)

print('Feature importances (XGBoost gain):')
print(feat_imp.to_string(index=False))
print()
print('Top signal: ' + feat_imp.iloc[0]['feature'])
print('This is the sensor driving risk scores most strongly.')

Feature importances (XGBoost gain):
            feature  importance
      SAR_change_dB    0.282266
        NDVI_change    0.239312
                 VV    0.093400
                 VH    0.087711
               NDVI    0.081753
               NDMI    0.077943
               NDWI    0.077565
dark_spot_magnitude    0.060050
     dark_spot_mask    0.000000
 chronic_spill_mask    0.000000
     new_spill_mask    0.000000

Top signal: SAR_change_dB
This is the sensor driving risk scores most strongly.


## Cell 6 — Isolation Forest anomaly detector

Isolation Forest complements XGBoost by finding **unusual signal combinations**
without relying on the rule-based labels.

It isolates anomalies by randomly partitioning feature space —
anomalies are isolated in fewer splits (short path length).

Contamination=0.1 means we expect ~10% of pixels to be anomalous.
This is a reasonable prior for the Niger Delta given known theft levels.

In [6]:
iso_forest = IsolationForest(
    n_estimators=200,
    contamination=0.10,
    random_state=42,
    n_jobs=-1
)

iso_forest.fit(X_scaled)

# Predictions: -1 = anomaly, 1 = normal
iso_pred   = iso_forest.predict(X_scaled)
iso_scores = iso_forest.score_samples(X_scaled)  # lower = more anomalous

df_clean = df_clean.copy()
df_clean["iso_anomaly"]    = (iso_pred == -1).astype(int)
df_clean["iso_score"]      = iso_scores
df_clean["iso_score_norm"] = MinMaxScaler().fit_transform(
    (-iso_scores).reshape(-1, 1)
).flatten()  # 0=normal, 1=most anomalous

n_anomalies = df_clean["iso_anomaly"].sum()
print("Isolation Forest results:")
print("  Total samples  : " + str(len(df_clean)))
print("  Anomalies      : " + str(n_anomalies)
      + "  (" + str(round(100*n_anomalies/len(df_clean), 1)) + "%)")
print()

# Cross-tab with XGBoost labels
cross = pd.crosstab(
    df_clean["risk_label"].map({0:"LOW",1:"MEDIUM",2:"HIGH"}),
    df_clean["iso_anomaly"].map({0:"Normal",1:"Anomaly"}),
    rownames=["XGBoost label"],
    colnames=["IsoForest"]
)
print("Agreement between XGBoost labels and Isolation Forest:")
print(cross)
print()
print("Interpretation: HIGH-risk pixels flagged as anomalies = strong theft signal")


Isolation Forest results:
  Total samples  : 270
  Anomalies      : 27  (10.0%)

Agreement between XGBoost labels and Isolation Forest:
IsoForest      Anomaly  Normal
XGBoost label                 
HIGH                 3       0
LOW                 13     201
MEDIUM              11      42

Interpretation: HIGH-risk pixels flagged as anomalies = strong theft signal


## Cell 7 — Combine scores into final risk index

In [7]:
# XGBoost probability for HIGH class
xgb_proba = xgb_model.predict_proba(X_scaled)
df_clean["xgb_prob_low"]    = xgb_proba[:, 0]
df_clean["xgb_prob_medium"] = xgb_proba[:, 1]
df_clean["xgb_prob_high"]   = xgb_proba[:, 2] if xgb_proba.shape[1] > 2 else 0
df_clean["xgb_pred"]        = xgb_model.predict(X_scaled)

# Combined risk index: weighted average of XGBoost HIGH probability + IsoForest score
# Weights: 0.6 XGBoost (supervised) + 0.4 IsoForest (unsupervised)
xgb_score_norm = MinMaxScaler().fit_transform(
    df_clean[["xgb_prob_high"]]
).flatten()

df_clean["combined_risk_score"] = (
    0.6 * xgb_score_norm +
    0.4 * df_clean["iso_score_norm"]
)

# Final risk tier
def risk_tier(score):
    if score >= 0.65:
        return "HIGH"
    elif score >= 0.35:
        return "MEDIUM"
    else:
        return "LOW"

df_clean["risk_tier"] = df_clean["combined_risk_score"].apply(risk_tier)

tier_counts = df_clean["risk_tier"].value_counts()
print("Final risk tier distribution:")
for tier in ["HIGH", "MEDIUM", "LOW"]:
    n   = tier_counts.get(tier, 0)
    pct = round(100 * n / len(df_clean), 1)
    print("  " + tier + ": " + str(n) + " pixels  (" + str(pct) + "%)")

print()
print("Top 10 highest-risk locations:")
top10 = df_clean.nlargest(10, "combined_risk_score")[
    ["lat","lon","combined_risk_score","risk_tier",
     "VV","dark_spot_mask","SAR_change_dB","new_spill_mask"]
].round(3)
print(top10.to_string(index=False))


Final risk tier distribution:
  HIGH: 2 pixels  (0.7%)
  MEDIUM: 1 pixels  (0.4%)
  LOW: 267 pixels  (98.9%)

Top 10 highest-risk locations:
  lat   lon  combined_risk_score risk_tier      VV  dark_spot_mask  SAR_change_dB  new_spill_mask
5.637 6.625                0.978      HIGH -15.264               1         -0.203               0
5.727 6.625                0.943      HIGH -11.826               1         -0.315               0
5.502 6.535                0.493    MEDIUM -15.383               1          1.529               0
5.682 6.625                0.324       LOW -10.286               0          0.338               0
5.098 6.805                0.312       LOW  -0.729               0          1.983               0
5.457 6.535                0.265       LOW  -8.289               0          0.146               0
5.547 6.535                0.251       LOW  -7.487               0          0.813               0
5.637 6.850                0.220       LOW  -4.025               0         

## Cell 8 — Risk score map

In [9]:
print('Risk tier distribution:')
print(df_clean['risk_tier'].value_counts().to_string())
print()
print('Top 10 highest-risk locations:')
top10 = df_clean.nlargest(10, 'combined_risk_score')[
    ['lat','lon','combined_risk_score','risk_tier','VV','dark_spot_mask','SAR_change_dB']
].round(3)
print(top10.to_string(index=False))

Risk tier distribution:
risk_tier
LOW       267
HIGH        2
MEDIUM      1

Top 10 highest-risk locations:
  lat   lon  combined_risk_score risk_tier      VV  dark_spot_mask  SAR_change_dB
5.637 6.625                0.978      HIGH -15.264               1         -0.203
5.727 6.625                0.943      HIGH -11.826               1         -0.315
5.502 6.535                0.493    MEDIUM -15.383               1          1.529
5.682 6.625                0.324       LOW -10.286               0          0.338
5.098 6.805                0.312       LOW  -0.729               0          1.983
5.457 6.535                0.265       LOW  -8.289               0          0.146
5.547 6.535                0.251       LOW  -7.487               0          0.813
5.637 6.850                0.220       LOW  -4.025               0          2.831
5.592 6.580                0.180       LOW  -8.957               0          0.180
5.502 7.074                0.179       LOW  -3.023               0      

## Cell 9 — Save models and scored dataset

In [10]:
print("=== Saving Models and Outputs ===")

# Save models
xgb_path    = os.path.join(MODEL_DIR, "xgb_risk_scorer.json")
scaler_path = os.path.join(MODEL_DIR, "feature_scaler.pkl")
iso_path    = os.path.join(MODEL_DIR, "isolation_forest.pkl")

xgb_model.save_model(xgb_path)
joblib.dump(scaler, scaler_path)
joblib.dump(iso_forest, iso_path)

print("Models saved:")
print("  " + xgb_path)
print("  " + scaler_path)
print("  " + iso_path)
print()

# Save scored dataset
scored_path = os.path.join(CACHE_DIR, "m3_risk_scored.csv")
df_clean.to_csv(scored_path, index=False)
print("Scored dataset -> " + scored_path
      + "  (" + str(round(os.path.getsize(scored_path)/1024, 1)) + " KB)")

# Save feature column list for dashboard
config = {
    "feature_cols":  FEATURE_COLS,
    "model_version": "1.0",
    "trained":       datetime.now().isoformat(),
    "n_samples":     len(df_clean),
    "cv_accuracy":   round(float(cv_scores.mean()), 3),
    "risk_tiers":    df_clean["risk_tier"].value_counts().to_dict(),
}
with open(os.path.join(CACHE_DIR, "m3_model_config.json"), "w") as f:
    json.dump(config, f, indent=2)
print("Model config   -> data/cached/m3_model_config.json")
print()

# Summary
print("=" * 45)
print("MODULE 3 COMPLETE")
print("=" * 45)
print("Samples scored : " + str(len(df_clean)))
print("CV accuracy    : " + str(round(cv_scores.mean(), 3)))
print("HIGH risk zones: " + str((df_clean["risk_tier"] == "HIGH").sum()))
print("Models saved to: data/models/")
print("Scored CSV      : data/cached/m3_risk_scored.csv")


=== Saving Models and Outputs ===
Models saved:
  ../data/models\xgb_risk_scorer.json
  ../data/models\feature_scaler.pkl
  ../data/models\isolation_forest.pkl

Scored dataset -> ../data/cached\m3_risk_scored.csv  (80.5 KB)
Model config   -> data/cached/m3_model_config.json

MODULE 3 COMPLETE
Samples scored : 270
CV accuracy    : 0.981
HIGH risk zones: 2
Models saved to: data/models/
Scored CSV      : data/cached/m3_risk_scored.csv


## Module 3 complete

| Output | File | Used in |
|--------|------|---------|
| Risk-scored dataset |  | Module 4 dashboard |
| XGBoost model |  | Module 4 dashboard |
| Isolation Forest |  | Module 4 dashboard |
| Feature scaler |  | Module 4 dashboard |
| Feature importance chart |  | README |
| Risk score map |  | README |

**Next: Module 4 — Risk Fusion Dashboard**
- Streamlit app reading all cached outputs
- Interactive Folium map with risk tiers, fire clusters, SAR change
- Alert table for NNPC operators
- Deploy to Streamlit Community Cloud for live public demo

# Visualisation deferred to Module 4 Streamlit dashboard (app.py)
# All plots available interactively at: https://pipelinewatch-ng.streamlit.app
# Matplotlib disabled in notebooks — memory constraints on resource-limited kernels